pkg/proc/fncall.go

```diff
@@ -77,21 +76,55 @@ type functionCallState struct {
 	panicvar *Variable
 }
 
-// CallFunction starts a debugger injected function call on the current thread of p.
-// See runtime.debugCallV1 in $GOROOT/src/runtime/asm_amd64.s for a
-// description of the protocol.
-func CallFunction(p Process, expr string, retLoadCfg *LoadConfig, checkEscape bool) error {
+type callContext struct {
+	p Process
+
+	// checkEscape is true if the escape check should be performed.
+	checkEscape bool
+
+	// retLoadCfg is the load configuration used to load return values
+	retLoadCfg LoadConfig
+
+	// Write to continueRequest to request a call to Continue from the
+	// debugger's main goroutine.
+	// Read from continueCompleted to wait for the target process to stop at
+	// one of the interaction point of the function call protocol.
+	// To signal that evaluation is completed a value will be written to
+	// continueRequest having cont == false and the return values in ret.
+	continueRequest   chan<- continueRequest
+	continueCompleted <-chan struct{}
+}
+
+type continueRequest struct {
+	cont bool
+	err  error
+	ret  *Variable
+}
+
+func (callCtx *callContext) doContinue() {
+	callCtx.continueRequest <- continueRequest{cont: true}
+	<-callCtx.continueCompleted
+}
+
+func (callCtx *callContext) doReturn(ret *Variable, err error) {
```

Comment:
doReturn tells the goroutine executing `Continue` that expression evaluation is finished and the user needs to see the return value, either `ret` or `err` need to be passed so that they can be returned to the user.

Reply:
Ack, makes sense.